### DISCUSSION TITLE
Tilting a screenshot image

### DISCUSSION THREAD

**Discussion by suntong** [ImageMagick/ImageMagick#2740] [Help] [answered]
I've seen some impressive images (but I can't find it any more) that tilts a screenshot somewhat like this:

![image](https://user-images.githubusercontent.com/422244/96372496-1b117580-1135-11eb-8fa1-fdcead1a1836.png)

Here is another example, by the same app author,

![image](https://www.glue.codes/images/browser-1166-be7e9628480cf0b026c189d1d72ee873.jpg)

But what I saw tilts the screenshot more than these -- it is like viewing the screen from LLS (lower-lefthand-side) to URS (upper-righthand-side). 

I know there needs some heavy computation involved (thus it is impressive). Is there any ready-made ImageMagick based cli tool for it (somewhat like the reverse to the `whiteboard` tool by fmw42 that I used before)? Thx

**Comment by fmw42** [comment, marked as answer]
Try my 3Drotate script

**Comment by suntong** [comment]
Found it http://www.fmwconcepts.com/imagemagick/3Drotate/index.php 
Thanks @frmw42

**Comment by suntong** [comment]
How to set the background to transparent, to fill where the output image is outside the area of the perspective of the input image? 

I tried `bgcolor=none`, but that turns the whole output to black. 

For `bgcolor=none`, I read from http://www.imagemagick.org/script/color.php

> There is also a color called 'none' that is fully transparent. This color is shorthand for rgba(0, 0, 0, 0.0).

**Reply by fmw42** [reply]
-virtual-pixel none

**Comment by suntong** [comment]
Hmm... could you elaborate a bit more please?

I see the USAGE is 
`3Drotate option=value infile outfile`

I.e., there is nowhere I can see I can put `-virtual-pixel`, which is confirmed when I tried it:

```sh
$ 3Drotate pan=30 auto=zc -virtual-pixel none mandril.jpg mandril-new.jpg

--- UNKNOWN OPTION ---
...
```

Can give exact command please? Thx

**Reply by fmw42** [reply]
Look at option vp and skycolor.  vp is the virtual-pixel option when outside the input image and skycolor is the color when the perspective looks above the horizon. Also set the bg option (background) to none

**Comment by suntong** [comment]
I got 

```
$ 3Drotate pan=30 auto=zc vp=none bgcolor=none mandril.jpg mandril-new.jpg

VP=none IS NOT A VALID VALUE
. . .
```

Can give an exact working command please (maybe not right away)?

**Reply by fmw42** [reply]
convert -list virtual-pixel
Background
Black
CheckerTile
Dither
Edge
Gray
HorizontalTile
HorizontalTileEdge
Mirror
None  <=====
Random
Tile
Transparent
VerticalTile
VerticalTileEdge
White

None should be valid.

I checked the script and left off "none", but you can use "transparent". 

I have now fixed the script. So download again. Check the version data at the top.

Note that JPG does not support transparency. So you need to use PNG or TIFF for output

`3Drotate pan=30 auto=zc vp=none bgcolor=none mandril3.jpg mandril3-new.png`

![mandril3-new](https://user-images.githubusercontent.com/13458218/97121450-95b13680-16db-11eb-8076-7023cbc3f8c2.png)

**Reply by suntong** [reply]
(now I just know I should better do "reply" )

Oh, thanks a lot. Working as expected now! 

thanks

**Comment by suntong** [comment]
Also there seems to be a bug with `3Drotate`'s help text. Here is what I got:


```
$ 3Drotate


3Drotate:


USAGE: 3Drotate option=value infile outfile
USAGE: 3Drotate option=value infile outfile
USAGE: 3Drotate [-h or -help]
USAGE: 3Drotate [-h or -help]


OPTIONS:  any one or more
OPTIONS:  any one or more


pan      value     rotation about image vertical centerline; 
pan      value     rotation about image vertical centerline; 
                   -180 to +180 (deg); default=0
                   -180 to +180 (deg); default=0
tilt     value     rotation about image horizontal centerline; 
tilt     value     rotation about image horizontal centerline; 
                   -180 to +180 (deg); default=0
                   -180 to +180 (deg); default=0
. . .
```

Every help text is repeated twice.

**Comment by fmw42** [comment]
The code I developed was on a Mac and works properly. But some Linux systems duplicate the messages. You can edit the code and try one of two potential fixes for your system. See my Pointers for Use on my home page.

Fred

**Comment by suntong** [comment]
sorry I have no idea how to fix the above two problems.

**Reply by fmw42** [reply]
For the double issue. My home page says:

On some Linux systems, there may be an issue with duplicate lines of documentation when using the -help option. If this is the case, try changing the following two lines in functions usage1() and usage2(), respectively, just below the commented section at the top of the script as follows (suggested by Stefan Vorkoetter and Ted Baker):

sed >&2 -e '1,/^####/d; /^###/g; /^#/!q; s/^#//; s/^ //; 4,$p' "$PROGDIR/$PROGNAME"
To:
sed -e '1,/^####$/d; /^###$/,$d; s/^#//; s/^ //' "$PROGDIR/$PROGNAME"
sed >&2 -e '1,/^####/d; /^######/g; /^#/!q; s/^#*//; s/^ //; 4,$p' "$PROGDIR/$PROGNAME"
To:
sed -e '1,/^####$/d; /^######$/,$d; s/^#*//; s/^ //' "$PROGDIR/$PROGNAME"

or

sed >&2 -n -e '1,/^####/d; /^###/g; /^#/!q; s/^#//; s/^ //; 4,$p' "$PROGDIR/$PROGNAME"
To:
sed -e '1,/^####$/d; /^###$/,$d; s/^#//; s/^ //' "$PROGDIR/$PROGNAME"
sed >&2 -n -e '1,/^####/d; /^######/g; /^#/!q; s/^#*//; s/^ //; 4,$p' "$PROGDIR/$PROGNAME"
To:
sed -e '1,/^####$/d; /^######$/,$d; s/^#*//; s/^ //' "$PROGDIR/$PROGNAME"


So the script is just a text file. So  you can edit it in some text editor.

**Reply by fmw42** [reply]
For the vp=none issue, I checked my script and forgot to put in none option. But you can do vp=transparent. I am adding none now and will post it shortly.

**Comment by suntong** [comment]
Thanks. As I tried `3Drotate pan=30 auto=zc vp=transparent bgcolor=none mandril.jpg mandril-new.jpg` as suggested but the background is still black not  transparent.

**Reply by fmw42** [reply]
See my last post. JPG does not support transparency. You will need to output to PNG or TIFF or the like which do support transparency.
